### ISSUE TITLE
Error while tuning LLaVA-Lighting

### ISSUE THREAD

**Issue by zzhanghub** [haotian-liu/LLaVA#102]
### When did you clone our code?

I cloned the code base after 5/1/23

### Describe the issue

Issue:

Command:
```
#!/bin/bash

WEIGHT_VERSION=1


# Visual instruction tuning (1 hour)
srun -p rdbp1_a100_80g -n1 -N 1 --gres=gpu:1 \
torchrun --nnodes=1 --nproc_per_node=1 --master_port=25001 \
    llava/train/train_mem.py \
    --model_name_or_path /mnt/lustre/share_data/zhangzhao2/VG/ckpt/llava/llava_v1/7B \
    --version $WEIGHT_VERSION \
    --data_path /mnt/lustre/share_data/zhangzhao2/VG/instruction_data/LLaVA-Instruct-150K/llava_instruct_80k.json \
    --image_folder /mnt/lustre/share_data/dongzhiwei1/coco2014/train2014 \
    --vision_tower /mnt/lustre/share_data/zhangzhao2/VG/ckpt/openai/clip-vit-large-patch14 \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end True \
    --bf16 True \
    --output_dir ./checkpoints \
    --num_train_epochs 1 \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 2 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 5000 \
    --save_total_limit 1 \
    --learning_rate 2e-5 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --fsdp "full_shard auto_wrap" \
    --fsdp_transformer_layer_cls_to_wrap 'LlamaDecoderLayer' \
    --model_max_length 2048 \
    --gradient_checkpointing True \
    --dataloader_num_workers 4 \
    --lazy_preprocess True \
    --report_to none
```

Log: 
```
Loading checkpoint shards: 100%|          | 2/2 [01:31<00:00, 45.96s/it]
WARNING:root:Loading data...
WARNING:root:Formatting inputs...Skip in lazy mode
/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/distributed/fsdp/_init_utils.py:295: UserWarning: FSDP is switching to use `NO_SHARD` instead of ShardingStrategy.FULL_SHARD since the world size is 1.
  warnings.warn(
  0%|          | 0/5000 [00:00<?, ?it/s]Traceback (most recent call last):
  File "/mnt/cache/zhangzhao2/codes/LLaVA/llava/train/train_mem.py", line 13, in <module>
    train()
  File "/mnt/cache/zhangzhao2/codes/LLaVA/llava/train/train.py", line 569, in train
    trainer.train()
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/transformers/trainer.py", line 1662, in train
    return inner_training_loop(
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/transformers/trainer.py", line 1927, in _inner_training_loop
    tr_loss_step = self.training_step(model, inputs)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/transformers/trainer.py", line 2699, in training_step
    loss = self.compute_loss(model, inputs)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/transformers/trainer.py", line 2731, in compute_loss
    outputs = model(**inputs)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py", line 748, in forward
    output = self._fsdp_wrapped_module(*args, **kwargs)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/mnt/cache/zhangzhao2/codes/LLaVA/llava/model/llava.py", line 218, in forward
    outputs = self.model(
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/mnt/cache/zhangzhao2/codes/LLaVA/llava/model/llava.py", line 126, in forward
    image_forward_outs = vision_tower(images, output_hidden_states=True)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/transformers/models/clip/modeling_clip.py", line 934, in forward
    return self.vision_model(
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/transformers/models/clip/modeling_clip.py", line 859, in forward
    hidden_states = self.embeddings(pixel_values)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/transformers/models/clip/modeling_clip.py", line 195, in forward
    patch_embeds = self.patch_embedding(pixel_values)  # shape = [*, width, grid, grid]
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/nn/modules/module.py", line 1501, in _call_impl
    return forward_call(*args, **kwargs)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/nn/modules/conv.py", line 463, in forward
    return self._conv_forward(input, self.weight, self.bias)
  File "/mnt/cache/zhangzhao2/anaconda3/envs/torch2/lib/python3.9/site-packages/torch/nn/modules/conv.py", line 459, in _conv_forward
    return F.conv2d(input, weight, bias, self.stride,
RuntimeError: GET was unable to find an engine to execute this computation
```

**Comment by haotian-liu** [comment]
Hi @zzhanghub, thanks for the interest in our work.

Similar to Vicuna, 13B model finetuning needs 8x A100, you may do that with 4x A100 (80G), but this is not fully verified.

You may also see discussion here: #68 
Thanks!

**Comment by zzhanghub** [comment]
> Hi @zzhanghub, thanks for the interest in our work.
> 
> Similar to Vicuna, 13B model finetuning needs 8x A100, you may do that with 4x A100 (80G), but this is not fully verified.
> 
> You may also see discussion here: #68 Thanks!

Thanks for replying😊, but this bug doesn't seem to be caused by OOM. I wonder if torch2.0 is the reason

**Comment by haotian-liu** [comment]
@zzhanghub I remember a similar issue when I tried with a single A100, and using more GPUs solves the issue.  I recommend trying with a 7B model (with a small batch size like 1 or 2), or with pretraining, which shall be okay running on a single A100 (80G).

PyTorch 2.0 may not be the issue, as I am also using PyTorch 2.0 :)

**Comment by zzhanghub** [comment]
> @zzhanghub I remember a similar issue when I tried with a single A100, and using more GPUs solves the issue. I recommend trying with a 7B model (with a small batch size like 1 or 2), or with pretraining, which shall be okay running on a single A100 (80G).
> 
> PyTorch 2.0 may not be the issue, as I am also using PyTorch 2.0 :)

@haotian-liu
Hi~ 
I was doing tuning on 7B. I tried two settings, both on 80g A100: one using 1 GPU and a small batch size, the other using 8 GPUs and a normal batch size. But I got the same error in both settings.

I also tried using the pretraining script, the script is as follows:
```bash
srun -p rdbp1_a100_80g -n1 -N 1 --gres=gpu:8 --comment wbsR-SC230944.001 \
torchrun --nnodes=1 --nproc_per_node=8 --master_port=25001 \
    llava/train/train_mem.py \
    --model_name_or_path /mnt/lustre/share_data/zhangzhao2/VG/ckpt/llava/llava_v1/7B \
    --version $WEIGHT_VERSION \
    --data_path /mnt/lustre/share_data/zhangzhao2/VG/instruction_data/LLaVA-Instruct-150K/llava_instruct_80k.json \
    --image_folder /mnt/lustre/share_data/dongzhiwei1/coco2014/train2014 \
    --vision_tower /mnt/lustre/share_data/zhangzhao2/VG/ckpt/openai/clip-vit-large-patch14 \
    --tune_mm_mlp_adapter True \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end \
    --bf16 True \
    --output_dir /mnt/lustre/share_data/zhangzhao2/VG/log/llava \
    --num_train_epochs 1 \
    --per_device_train_batch_size 8 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 1 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 2400 \
    --save_total_limit 1 \
    --learning_rate 2e-3 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --model_max_length 2048 \
    --gradient_checkpointing True \
    --dataloader_num_workers 4 \
    --lazy_preprocess True \
    --report_to none
```
But I got an error:

![iShot_2023-05-06_09 49 12](https://user-images.githubusercontent.com/25194962/236591852-d72eae3f-15fc-418f-b6a0-bcded8cbe45f.png)

 I set `model_name_or_pat to 'llava' instead of 'vicuna'. I'm wondering if this could be the cause of the problem?

**Comment by haotian-liu** [comment]
@zzhanghub
This is strange, the error indicates that you used FSDP, but it does not appear in your command.

You used `--tune_mm_mlp_adapter True`, which only tunes the projection layer, and currently unless you install PyTorch nightly and follow instructions [here](https://github.com/haotian-liu/LLaVA#experimental-use-fsdp-to-save-memory-in-pretraining), it is not possible to use FSDP with partially frozen parameters, as PEFT is not supported fully yet.  You can only use FSDP in instruction tuning.

I am not sure if you enabled FSDP somewhere else?

And I also see that you are using instruction tuning dataset, so could you please confirm the command that you are using, and the intent of the experiment?  This may let me better understand the issue, thanks.

**Comment by zzhanghub** [comment]
I haven't enabled FSDP anywhere else because I haven't made any modifications to the llava code yet.

The purpose of me conducting this experiment is to further tune the llava-ckpt based on a new dataset. However, I haven't used the new dataset yet, and instead I used the data you provided to test my code.

**Comment by haotian-liu** [comment]
I see, if so, can you try printing `training_args.fsdp` and see what's configured?  It should not be enabled by default, but we can see what's happening behind.  You may check this right after this:

```
    parser = transformers.HfArgumentParser(
        (ModelArguments, DataArguments, TrainingArguments))
    model_args, data_args, training_args = parser.parse_args_into_dataclasses()
```

**Comment by haotian-liu** [comment]
Also as a direct solution (if you are not planning to tune the projection layer only), you may try removing `--tune_mm_mlp_adapter True` and running with 8 GPUs, this will work with FSDP enabled.

**Comment by zzhanghub** [comment]
@haotian-liu 
I printed out the `training_args.fsdp` and the result is as follows:
![image](https://user-images.githubusercontent.com/25194962/236596561-e77f7c65-88c7-4ed4-a825-aa46432387f1.png)

I removed `--tune_mm_mlp_adapter Tru` and use 8 GPUs for pre-training, and encountered a similar problem as before
``` shell
srun -p a100_80g -n1 -N 1 --gres=gpu:8  \
torchrun --nnodes=1 --nproc_per_node=8 --master_port=25001 \
    llava/train/train_mem.py \
    --model_name_or_path /mnt/lustre/share_data/zhangzhao2/VG/ckpt/llava/llava_v1/7B \
    --version $WEIGHT_VERSION \
    --data_path /mnt/lustre/share_data/zhangzhao2/VG/instruction_data/LLaVA-Instruct-150K/llava_instruct_80k.json \
    --image_folder /mnt/lustre/share_data/dongzhiwei1/coco2014/train2014 \
    --vision_tower /mnt/lustre/share_data/zhangzhao2/VG/ckpt/openai/clip-vit-large-patch14 \
    --mm_vision_select_layer -2 \
    --mm_use_im_start_end \
    --bf16 True \
    --output_dir /mnt/lustre/share_data/zhangzhao2/VG/log/llava \
    --num_train_epochs 1 \
    --per_device_train_batch_size 1 \
    --per_device_eval_batch_size 4 \
    --gradient_accumulation_steps 1 \
    --evaluation_strategy "no" \
    --save_strategy "steps" \
    --save_steps 2400 \
    --save_total_limit 1 \
    --learning_rate 2e-3 \
    --weight_decay 0. \
    --warmup_ratio 0.03 \
    --lr_scheduler_type "cosine" \
    --logging_steps 1 \
    --tf32 True \
    --model_max_length 2048 \
    --gradient_checkpointing True \
    --dataloader_num_workers 4 \
    --lazy_preprocess True \
    --report_to none
```

![image](https://user-images.githubusercontent.com/25194962/236595719-ece495d2-56b5-4597-a636-0e8b39c6e6e3.png)

**Comment by haotian-liu** [comment]
Thanks for providing this info.  I am wondering how did you setup your slurm cuda environment?  Is it a new environment?  Furthermore, what's your PyTorch and CUDA version?  Thanks.

**Comment by zzhanghub** [comment]
@haotian-liu 
Here are some version information :)
![image](https://user-images.githubusercontent.com/25194962/236597951-f409da39-61f7-41d4-82ab-6f268241f8a8.png)

![image](https://user-images.githubusercontent.com/25194962/236597787-f7a69c03-3a88-4c4c-bba9-0deabd609485.png)
![image](https://user-images.githubusercontent.com/25194962/236597816-38291bcc-5a75-4d7e-b1d7-087ad740a6ba.png)
![image](https://user-images.githubusercontent.com/25194962/236597836-499598d8-2a96-4ca0-8695-91df4dfe9cd5.png)

**Comment by haotian-liu** [comment]
What's your `transformers` version?  And can you also please confirm that this is a new environment for LLaVA only?

**Comment by zzhanghub** [comment]
> What's your `transformers` version? And can you also please confirm that this is a new environment for LLaVA only?.
Yes, it is a new environment for LLaVA.
![image](https://user-images.githubusercontent.com/25194962/236600931-01da427c-2386-4de7-8663-c3f4d275319d.png)
@haotian-liu

**Comment by haotian-liu** [comment]
This may be the issue as the `transformers` version is not correct, as it should show something like `dev`.  Please try installing the correct version of `transformers` following the instructions [here](https://github.com/haotian-liu/LLaVA#upgrade-to-v01).

**Comment by zzhanghub** [comment]
> This may be the issue as the `transformers` version is not correct, as it should show something like `dev`. Please try installing the correct version of `transformers` following the instructions [here](https://github.com/haotian-liu/LLaVA#upgrade-to-v01).

@haotian-liu 
<img width="275" alt="image" src="https://user-images.githubusercontent.com/25194962/236668427-98f7526f-3d56-4391-9388-ea48081f79bf.png">
I have created a new environment again, but the situation seems to have not changed.

**Comment by zzhanghub** [comment]
> This may be the issue as the `transformers` version is not correct, as it should show something like `dev`. Please try installing the correct version of `transformers` following the instructions [here](https://github.com/haotian-liu/LLaVA#upgrade-to-v01).

@haotian-liu 
Good news! I finally succeeded in downgrading the Torch version to **1.13.1** and CUDA=**11.7**!

I spent a lot of effort building `flash-attn`, and I know that `torch` and `flash-attn` require consistent use of CUDA. However, I failed to compile them with many different CUDA versions and GCC versions (the versions of `torch` and `flash-attn` are corresponding). Finally, I succeeded with `CUDA=11.7` and `gcc=7.3.0`.

Regarding the bug I encountered before, I think it may be related to what is discussed in this (issue)[https://github.com/pytorch/pytorch/issues/92907], and it should not be due to `LLaVA` or `transformers`. However, I don't know why the error occurred since many people can use `torch-2.0` without any problems.

Lastly, I would like to express my gratitude for your active replies and support. Your assistance gave me confidence that LLaVA will be a reliable and promising project!
